### PR TITLE
Add profile-guided optimisation option to build script

### LINKED ISSUE #556: Profile-guided optimization (PGO) as optional part of builds
The `pydantic` project had a bit of success adding PGO to their builds, as shown in the relevant PRs:
- https://github.com/pydantic/pydantic-core/pull/678
- https://github.com/pydantic/pydantic-core/pull/1063
- https://github.com/pydantic/pydantic-core/pull/741
- 
I'm not sure if we'd see similar performance improvements from adding this to our build chain, but a 10-20% improvement might be doable and would not add much to actual release builds.

Let's investigate.

### PR DESCRIPTION
**Pull request by N-Wouda**
This PR:

- [x] Closes #556.
- [ ] Adds and passes relevant tests.
- [ ] Has well-formatted code and documentation.

### Benchmark on VRPTW

This PR (47f0dc2353b9d540743a124b53fe81dc4c537aad; job ID 7853551):
```
                MEAN       MIN       MAX  COV
      inst
obj   C1     41370.5   41337.2   41410.1  0.1
      C2     16201.1   16197.0   16204.4  0.0
      R1     47275.8   47227.9   47369.6  0.1
      R2     27810.9   27769.4   27856.3  0.1
      RC1    44286.4   44252.1   44336.9  0.1
      RC2    23294.8   23278.0   23331.1  0.1
iters C1     76502.0   60675.2   82545.9  7.9
      C2    112465.3  101503.5  117537.5  4.2
      R1     40315.5   33021.9   43240.8  7.6
      R2     70731.8   60786.7   76122.6  7.3
      RC1    41620.9   33969.7   44552.4  7.6
      RC2    78788.7   66588.6   85885.9  7.6
```

Current main (2feef56c67477d3e7e69ae7da6617962f53ad3ba; job ID 7852597):
```
                MEAN      MIN       MAX   COV
      inst
obj   C1     41372.7  41338.2   41409.0   0.1
      C2     16201.1  16197.4   16204.2   0.0
      R1     47269.2  47240.2   47317.9   0.0
      R2     27813.0  27767.3   27859.4   0.1
      RC1    44288.1  44258.0   44354.1   0.1
      RC2    23293.7  23277.6   23323.8   0.1
iters C1     77582.0  66226.6   84924.5   7.1
      C2    112501.9  96835.1  137578.3   9.4
      R1     41034.4  37063.8   47525.9   7.4
      R2     73104.5  59404.6   88665.7  11.5
      RC1    41734.7  35304.5   48645.4   9.5
      RC2    75931.1  61890.8   91481.5  10.0
```

<details>

**Notes**:

- It is essential that you add a test when making code changes.
  This keeps the code coverage level up, and helps ensure the changes work as intended.
- When fixing a bug, you must add a test that would produce the bug in the master branch, and then show that it is fixed with the new code. 
- New code additions must be well formatted. Changes should pass the pre-commit workflow, which you can set up locally using [pre-commit](https://pre-commit.com/#intro). 
- Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

**Comment by N-Wouda** [comment]
I'm not sure if the resulting binaries are actually better/faster, but that's for a benchmark to decide. The current imlementation should be functional, but needs clean up.

File: build_extensions.py [hunk 1]
```diff
@@ -122,14 +127,39 @@ def main():
...
+        check_call(cmd.split(" "))
+
+        configure(
+            build_dir,
+            args.build_type,
+            args.problem,
+            [*args.additional, "-Db_pgo=use"],
+        )
+        compile(build_dir, args.verbose)
+        install(build_dir)
```
Comment (N-Wouda) [inline build_extensions.py:153]: The idea is:
- Generate build with PGO collection set up
- Do some representative work to collect data
- Use that data in another build to further optimise for hot spots.

File: build_extensions.py [hunk 2]
```diff
@@ -62,28 +59,11 @@ def clean(build_dir: pathlib.Path, install_dir: pathlib.Path):
...
-        check_call(
-            ["stubgen", "--parse-only", "-o", ".", "-m", ext_name],
-            cwd=ext_dir,
-        )
-
-    for extension in install_dir.rglob("*.so"):
-        regen(extension)
-
-    for extension in install_dir.rglob("*.pyd"):
-        regen(extension)
```
Comment (N-Wouda) [inline build_extensions.py:79]: We don't use this at all so it's time to go.

**Comment by N-Wouda** [comment]
Some local tinkering suggests it makes a few percent of difference (say, 5%?), but I don't have a full picture yet. I'll run a benchmark with this tomorrow.

**Comment by N-Wouda** [comment]
OK so this - at least with the currently chosen workload to profile - has no meaningful impact. At least now we know.

**Review by leonlan — APPROVED** [review]

**Pull request merged**